source/Halibut/Queue/Redis/MessageStorage/MessageSerialiserAndDataStreamStorageWithExceptionObserver.cs

```diff
@@ -0,0 +1,79 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Halibut.Queue.QueuedDataStreams;
+using Halibut.Queue.Redis.RedisHelpers;
+using Halibut.Transport.Protocol;
+
+namespace Halibut.Queue.Redis.MessageStorage
+{
+    /// <summary>
+    /// Decorator implementation of IMessageSerialiserAndDataStreamStorage that wraps another implementation
+    /// and notifies an observer when exceptions occur during any of the operations.
+    /// </summary>
+    public class MessageSerialiserAndDataStreamStorageWithExceptionObserver : IMessageSerialiserAndDataStreamStorage
+    {
+        readonly IMessageSerialiserAndDataStreamStorage inner;
+        readonly IMessageSerialiserAndDataStreamStorageExceptionObserver exceptionObserver;
+
+        public MessageSerialiserAndDataStreamStorageWithExceptionObserver(
+            IMessageSerialiserAndDataStreamStorage inner, 
+            IMessageSerialiserAndDataStreamStorageExceptionObserver exceptionObserver)
+        {
+            this.inner = inner ?? throw new ArgumentNullException(nameof(inner));
```

Comment:
Why are we guarding for null? This isn't a pattern we tend to do at Octopus, as we tend to use the compiler for this.

Reply:
I guess we mostly don't really need to as the type system mostly guards against this although not in all cases I think we might still have some null checks off in halibut I can't remember. In any case "the ai wrote this, and I don't think it hurts to have it."